### PR TITLE
Restore collection.Map --, Fixes scala/bug#10957

### PR DESCRIPTION
**Pull request by MasseGuillaume**

### CODE REVIEW

File: src/library/scala/collection/Map.scala [hunk 1]
```diff
@@ -288,11 +288,18 @@ trait MapOps[K, +V, +CC[_, _] <: IterableOps[_, AnyConstr, _], +C]
...
+  def + [V1 >: V](kv: (K, V1)): CC[K, V1] =
+    mapFactory.from(new View.Appended(toIterable, kv))
 
   @deprecated("Use ++ with an explicit collection argument instead of + with varargs", "2.13.0")
-  def + [V1 >: V](elem1: (K, V1), elem2: (K, V1), elems: (K, V1)*): CC[K, V1] = mapFactory.from(new View.Concat(new View.Appended(new View.Appended(toIterable, elem1), elem2), elems))
+  def + [V1 >: V](elem1: (K, V1), elem2: (K, V1), elems: (K, V1)*): CC[K, V1] =
+    mapFactory.from(new View.Concat(new View.Appended(new View.Appended(toIterable, elem1), elem2), elems))
+
+  @deprecated("Consider requiring an immutable Map.", "2.13.0")
+  def -- [V1 >: V](that: IterableOnce[K]): CC[K, V1] = {
```
Comment (julienrf) [inline src/library/scala/collection/Map.scala]: The type parameter `V1` is not needed, and we should return a `C`. (You can see in `immutable.MapOps` the right signature to use)

File: src/library/scala/collection/Map.scala [hunk 2]
```diff
@@ -288,11 +288,18 @@ trait MapOps[K, +V, +CC[_, _] <: IterableOps[_, AnyConstr, _], +C]
...
+    mapFactory.from(new MapView.SubtractAll(this, that))
```
Comment (julienrf) [inline src/library/scala/collection/Map.scala]: We can just use `filterKeys` rather than implementing a new `MapView`.

**Review by julienrf — CHANGES_REQUESTED** [review]

File: src/library/scala/collection/Map.scala (hunk 2 above)
Comment (MasseGuillaume) [reply src/library/scala/collection/Map.scala]: Ok, it looks like I need to keep the `V1 >: V`

```
[error]  found   : MapOps.this.MapCC[K,V]
[error]     (which expands to)  CC[K,V]
[error]  required: C
[error]     mapFactory.from(this.filterKeys(k => !keysSet.contains(k)))
[error]                    ^
```

Comment (julienrf) [reply src/library/scala/collection/Map.scala]: You should use `fromSpecificIterable` instead of `mapFactory.from`.

Comment (MasseGuillaume) [reply src/library/scala/collection/Map.scala]: Now when I run the unit test I get this:

```
[error] Test scala.collection.MapTest.test failed: java.lang.VerifyError: class scala.collection.immutable.HashMap overrides final method $minus$minus.(Lscala/collection/IterableOnce;)Ljava/lang/Object;, took 0.074 sec
[error]     at java.lang.ClassLoader.defineClass1(Native Method)
[error]     at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
[error]     at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
[error]     at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
[error]     at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
[error]     at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
[error]     at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
[error]     at java.security.AccessController.doPrivileged(Native Method)
[error]     at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
[error]     at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
[error]     at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:338)
[error]     at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
[error]     at scala.collection.immutable.Map$Map3.updated(Map.scala:266)
[error]     at scala.collection.immutable.Map$Map3.updated(Map.scala:242)
[error]     at scala.collection.immutable.MapOps.$plus(Map.scala:99)
[error]     at scala.collection.immutable.MapOps.$plus$(Map.scala:99)
[error]     at scala.collection.immutable.AbstractMap.$plus(Map.scala:325)
[error]     at scala.collection.immutable.Map$$anon$1.addOne(Map.scala:177)
[error]     at scala.collection.immutable.Map$$anon$1.addOne(Map.scala:176)
[error]     at scala.collection.mutable.Growable.addAll(Growable.scala:49)
[error]     at scala.collection.mutable.Growable.addAll$(Growable.scala:46)
[error]     at scala.collection.mutable.ImmutableBuilder.addAll(ImmutableBuilder.scala:9)
[error]     at scala.collection.mutable.Growable.$plus$plus$eq(Growable.scala:55)
[error]     at scala.collection.mutable.Growable.$plus$plus$eq$(Growable.scala:55)
[error]     at scala.collection.mutable.ImmutableBuilder.$plus$plus$eq(ImmutableBuilder.scala:9)
[error]     at scala.collection.immutable.Map$.from(Map.scala:172)
[error]     at scala.collection.immutable.Map$.from(Map.scala:132)
[error]     at scala.collection.MapFactory$Delegate.from(Factory.scala:403)
[error]     at scala.collection.MapFactory.apply(Factory.scala:369)
[error]     at scala.collection.MapFactory.apply$(Factory.scala:369)
[error]     at scala.collection.MapFactory$Delegate.apply(Factory.scala:402)
[error]     at scala.collection.MapTest.test(MapTest.scala:15)
[error]     ...
```

Comment (julienrf) [reply src/library/scala/collection/Map.scala]: Is `--` final in `collection.MapOps`? It should not. It should be overridden by `collection.immutable.MapOps`.

Comment (MasseGuillaume) [reply src/library/scala/collection/Map.scala]: I added another commit so you can see the failure:
https://github.com/scala/scala/pull/6915/commits/e3ab7f5c8cfc541d0ddf654a246b3f47e8307c04

It's not final in `collection.MapOps`, It's overridden in `immutable.MapOps`.

Comment (julienrf) [reply src/library/scala/collection/Map.scala]: Oh, it’s scala/bug#10853 again, I think.

Replace `final` with:

~~~ scala
@deprecatedOverriding("This method should be final, but is not due to scala/bug#10853", "2.13.0")
~~~

Comment (MasseGuillaume) [reply src/library/scala/collection/Map.scala]: Ok, it works now. I just squashed the commits, let's wait for the ci.

**Review by julienrf — APPROVED** [review]

**Pull request merged**